tests/components/esphome/test_config_flow.py

```diff
@@ -245,16 +289,41 @@ async def test_discovery_duplicate_data(hass, mock_client):
         "properties": {"address": "test8266.local"},
     }
 
-    mock_client.device_info.return_value = mock_coro(MockDeviceInfo(False, "test8266"))
+    mock_client.device_info = AsyncMock(return_value=MockDeviceInfo(False, "test8266"))
 
     result = await hass.config_entries.flow.async_init(
         "esphome", data=service_info, context={"source": "zeroconf"}
     )
-    assert result["type"] == "form"
+    assert result["type"] == RESULT_TYPE_FORM
     assert result["step_id"] == "discovery_confirm"
 
     result = await hass.config_entries.flow.async_init(
         "esphome", data=service_info, context={"source": "zeroconf"}
     )
-    assert result["type"] == "abort"
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_in_progress"
+
+
+async def test_discovery_updates_unique_id(hass, mock_client):
+    """Test a duplicate discovery host aborts and updates existing entry."""
+    entry = MockConfigEntry(
+        domain="esphome",
+        data={CONF_HOST: "192.168.43.183", CONF_PORT: 6053, CONF_PASSWORD: ""},
+    )
+
+    entry.add_to_hass(hass)
+
+    service_info = {
+        "host": "192.168.43.183",
+        "port": 6053,
+        "hostname": "test8266.local.",
+        "properties": {"address": "test8266.local"},
+    }
+    result = await hass.config_entries.flow.async_init(
+        "esphome", context={"source": "zeroconf"}, data=service_info
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5451397)